### PR TITLE
Smoke Timer

### PR DESCRIPTION
**Pull request by blackMasterLord**
https://user-images.githubusercontent.com/76704816/128630454-848459fb-8cbd-4072-abaa-99d0069d6660.mov

### CODE REVIEW

**Comment by KisSsArt** [comment]
screenshot?

**Comment by blackMasterLord** [comment]
> screenshot?

done

**Comment by KisSsArt** [comment]
cool!

**Comment by floydGG** [comment]
Big thanks! yeah anyway, how to change the background alpha?

**Comment by tusharborkar360** [comment]
Thank You so much! It's helpful!

**Comment by n0verify** [comment]
> Big thanks! yeah anyway, how to change the background alpha?

in line
 -->       drawList->AddRectFilled(rect_out.Min, rect_out.Max, IM_COL32_WHITE);
replace IM_COL32_WHITE
to remove bg, i made mine like this
-->       drawList->AddRectFilled(rect_out.Min, rect_out.Max, IM_COL32_BLACK_TRANS);

**Comment by blackMasterLord** [comment]
> > Big thanks! yeah anyway, how to change the background alpha?
> 
> in line
> --> drawList->AddRectFilled(rect_out.Min, rect_out.Max, IM_COL32_WHITE);
> replace IM_COL32_WHITE
> to remove bg, i made mine like this
> --> drawList->AddRectFilled(rect_out.Min, rect_out.Max, IM_COL32_BLACK_TRANS);

instead of `IM_COL32_WHITE` u can use `IM_COL32(R, G, B, A)` to setting color;

**Comment by floydGG** [comment]
> > > Big thanks! yeah anyway, how to change the background alpha?
> > 
> > 
> > in line
> > --> drawList->AddRectFilled(rect_out.Min, rect_out.Max, IM_COL32_WHITE);
> > replace IM_COL32_WHITE
> > to remove bg, i made mine like this
> > --> drawList->AddRectFilled(rect_out.Min, rect_out.Max, IM_COL32_BLACK_TRANS);
> 
> instead of `IM_COL32_WHITE` u can use `IM_COL32(R, G, B, A)` to setting color;

what about the timer color? black is kinda hard to see

**Comment by blackMasterLord** [comment]
> > > > Big thanks! yeah anyway, how to change the background alpha?
> > > 
> > > 
> > > in line
> > > --> drawList->AddRectFilled(rect_out.Min, rect_out.Max, IM_COL32_WHITE);
> > > replace IM_COL32_WHITE
> > > to remove bg, i made mine like this
> > > --> drawList->AddRectFilled(rect_out.Min, rect_out.Max, IM_COL32_BLACK_TRANS);
> > 
> > 
> > instead of `IM_COL32_WHITE` u can use `IM_COL32(R, G, B, A)` to setting color;
> 
> what about the timer color? black is kinda hard to see

`drawList->AddText({ pos.x - (textSize.x / 2), pos.y - (textSize.y / 2) }, IM_COL32_BLACK, text.str().c_str());`
change `IM_COL32_BLACK `to `IM_COL32(R, G, B, A)`

**Comment by blackMasterLord** [comment]
> > > > Big thanks! yeah anyway, how to change the background alpha?
> > > 
> > > 
> > > in line
> > > --> drawList->AddRectFilled(rect_out.Min, rect_out.Max, IM_COL32_WHITE);
> > > replace IM_COL32_WHITE
> > > to remove bg, i made mine like this
> > > --> drawList->AddRectFilled(rect_out.Min, rect_out.Max, IM_COL32_BLACK_TRANS);
> > 
> > 
> > instead of `IM_COL32_WHITE` u can use `IM_COL32(R, G, B, A)` to setting color;
> 
> what about the timer color? black is kinda hard to see

**now u can change color in game**

**Comment by floydGG** [comment]
> > > > > Big thanks! yeah anyway, how to change the background alpha?
> > > > 
> > > > 
> > > > in line
> > > > --> drawList->AddRectFilled(rect_out.Min, rect_out.Max, IM_COL32_WHITE);
> > > > replace IM_COL32_WHITE
> > > > to remove bg, i made mine like this
> > > > --> drawList->AddRectFilled(rect_out.Min, rect_out.Max, IM_COL32_BLACK_TRANS);
> > > 
> > > 
> > > instead of `IM_COL32_WHITE` u can use `IM_COL32(R, G, B, A)` to setting color;
> > 
> > 
> > what about the timer color? black is kinda hard to see
> 
> **now u can change color in game**

big thanks!

**Comment by Asviix** [comment]
damn that's really good work, well done

File: Osiris/Hacks/Visuals.cpp [hunk 1]
```diff
@@ -68,6 +68,10 @@ struct VisualsConfig {
     float hitMarkerTime{ 0.6f };
     BulletTracers bulletTracers;
     ColorToggle molotovHull{ 1.0f, 0.27f, 0.0f, 0.3f };
+    bool smokeTimer{ false };
+    Color4 smokeTimer_BG{ 1.0f, 1.0f, 1.0f, 0.5f };
+    Color4 smokeTimer_TIMER{ 0.0f, 0.0f, 1.0f, 1.0f };
+    Color4 smokeTimer_TEXT{ 0.0f, 0.0f, 0.0f, 1.0f };
```
Comment (danielkrupinski) [inline Osiris/Hacks/Visuals.cpp:74]: Create a struct that will hold those

File: Osiris/Hacks/Visuals.cpp [hunk 2]
```diff
@@ -701,6 +713,69 @@ void Visuals::drawMolotovHull(ImDrawList* drawList) noexcept
...
+static std::vector<smokeData> smokes;
+
+void Visuals::drawSmokeTimerEvent(GameEvent* event) noexcept
+{
+    if (event) {
+        smokeData data{};
+        auto time = memory->globalVars->realtime + SMOKEGRENADE_LIFETIME;
+        auto pos = Vector(event->getFloat("x"), event->getFloat("y"), event->getFloat("z"));
+        data.destructionTime = time;
+        data.pos = pos;
```
Comment (danielkrupinski) [inline Osiris/Hacks/Visuals.cpp:733]: Unnecessary introducing `time` and `pos` variables.

File: Osiris/Hacks/Visuals.cpp [hunk 3]
```diff
@@ -701,6 +713,69 @@ void Visuals::drawMolotovHull(ImDrawList* drawList) noexcept
...
+    float destructionTime;
+    Vector pos;
+};
+
+static std::vector<smokeData> smokes;
+
+void Visuals::drawSmokeTimerEvent(GameEvent* event) noexcept
+{
+    if (event) {
+        smokeData data{};
```
Comment (danielkrupinski) [inline Osiris/Hacks/Visuals.cpp:729]: Unnecessary zeroing memory that is overwritten a moment later.

**Comment by Yoka1998** [comment]
Bro can you help me?
I can't open cheat menu in game, do you know any fix for that?
If so, please let me know.
Thanks

**Comment by Yoka1998** [comment]
Please help me i can't close menu in game after i coded smoke timer :(

**Pull request closed without merging**